File: app/golang/internal/order/infrastructure/web/order_controller.go
package web

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
	orderApp "github.com/orkhan/ecommerce/internal/order/application"
	orderDomain "github.com/orkhan/ecommerce/internal/order/domain"
	"github.com/orkhan/ecommerce/internal/shared/application/bus"
	"github.com/orkhan/ecommerce/internal/shared/infrastructure/api"
)

type OrderController struct {
	cmdBus   *bus.Bus
	queryBus *bus.QueryHandlers
}

func NewOrderController(cmdBus *bus.Bus, queryBus *bus.QueryHandlers) *OrderController {
	return &OrderController{cmdBus: cmdBus, queryBus: queryBus}
}

func (c *OrderController) Store(ctx *gin.Context) {
	var cmd orderApp.CreateOrderCommand
	if err := ctx.ShouldBindJSON(&cmd); err != nil {
		ctx.Error(err)
		return
	}
	id, err := bus.Dispatch[orderApp.CreateOrderCommand, uuid.UUID](
		ctx.Request.Context(), c.cmdBus, cmd)
	if err != nil {
		ctx.Error(err)
		return
	}
	ctx.JSON(http.StatusCreated, api.SuccessWithMessage(gin.H{"id": id}, "Sifariş yaradıldı"))
}

func (c *OrderController) Show(ctx *gin.Context) {
	id, err := uuid.Parse(ctx.Param("id"))
	if err != nil {
		ctx.Error(err)
		return
	}
	dto, err := bus.Ask[orderApp.GetOrderQuery, orderApp.OrderDTO](
		ctx.Request.Context(), c.queryBus, orderApp.GetOrderQuery{OrderID: id})
	if err != nil {
		ctx.Error(err)
		return
	}
	ctx.JSON(http.StatusOK, api.Success(dto))
}

func (c *OrderController) ListByUser(ctx *gin.Context) {
	id, err := uuid.Parse(ctx.Param("userId"))
	if err != nil {
		ctx.Error(err)
		return
	}
	dtos, err := bus.Ask[orderApp.ListOrdersQuery, []orderApp.OrderDTO](
		ctx.Request.Context(), c.queryBus, orderApp.ListOrdersQuery{UserID: id})
	if err != nil {
		ctx.Error(err)
		return
	}
	ctx.JSON(http.StatusOK, api.Success(dtos))
}

func (c *OrderController) Cancel(ctx *gin.Context) {
	id, err := uuid.Parse(ctx.Param("id"))
	if err != nil {
		ctx.Error(err)
		return
	}
	var body struct {
		Reason string `json:"reason"`
	}
	_ = ctx.ShouldBindJSON(&body)

	_, err = bus.Dispatch[orderApp.CancelOrderCommand, struct{}](
		ctx.Request.Context(), c.cmdBus,
		orderApp.CancelOrderCommand{OrderID: id, Reason: body.Reason})
	if err != nil {
		ctx.Error(err)
		return
	}
	ctx.JSON(http.StatusOK, api.SuccessWithMessage(nil, "Sifariş ləğv edildi"))
}

func (c *OrderController) UpdateStatus(ctx *gin.Context) {
	id, err := uuid.Parse(ctx.Param("id"))
	if err != nil {
		ctx.Error(err)
		return
	}
	var body struct {
		Target string `json:"target" binding:"required"`
	}
	if err := ctx.ShouldBindJSON(&body); err != nil {
		ctx.Error(err)
		return
	}

	_, err = bus.Dispatch[orderApp.UpdateOrderStatusCommand, struct{}](
		ctx.Request.Context(), c.cmdBus,
		orderApp.UpdateOrderStatusCommand{
			OrderID: id, Target: orderDomain.OrderStatus(body.Target),
		})
	if err != nil {
		ctx.Error(err)
		return
	}
	ctx.JSON(http.StatusOK, api.SuccessWithMessage(nil, "Status yeniləndi"))
}

func (c *OrderController) RegisterRoutes(authed *gin.RouterGroup) {
	authed.POST("/orders", c.Store)
	authed.GET("/orders/:id", c.Show)
	authed.GET("/orders/user/:userId", c.ListByUser)
	authed.POST("/orders/:id/cancel", c.Cancel)
	authed.PATCH("/orders/:id/status", c.UpdateStatus)
}

File: app/spring/src/main/java/az/ecommerce/order/infrastructure/web/OrderController.java
package az.ecommerce.order.infrastructure.web;

import az.ecommerce.order.application.command.cancelorder.CancelOrderCommand;
import az.ecommerce.order.application.command.createorder.CreateOrderCommand;
import az.ecommerce.order.application.command.updateorderstatus.UpdateOrderStatusCommand;
import az.ecommerce.order.application.dto.OrderDto;
import az.ecommerce.order.application.query.getorder.GetOrderQuery;
import az.ecommerce.order.application.query.listorders.ListOrdersQuery;
import az.ecommerce.order.domain.enums.OrderStatusEnum;
import az.ecommerce.shared.application.bus.CommandBus;
import az.ecommerce.shared.application.bus.QueryBus;
import az.ecommerce.shared.infrastructure.api.ApiResponse;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Laravel: app/Http/Controllers/OrderController.php
 *   POST   /api/orders              — auth + throttle:orders
 *   GET    /api/orders/{id}         — auth
 *   GET    /api/orders/user/{id}    — auth
 *   POST   /api/orders/{id}/cancel  — auth + policy
 *   PATCH  /api/orders/{id}/status  — auth + policy
 */
@RestController
@RequestMapping("/api/orders")
@PreAuthorize("isAuthenticated()")
public class OrderController {

    private final CommandBus commandBus;
    private final QueryBus queryBus;

    public OrderController(CommandBus commandBus, QueryBus queryBus) {
        this.commandBus = commandBus;
        this.queryBus = queryBus;
    }

    @PostMapping
    public ResponseEntity<ApiResponse<Map<String, UUID>>> store(@RequestBody @Valid CreateOrderCommand cmd) {
        UUID id = commandBus.dispatch(cmd);
        return ResponseEntity.status(HttpStatus.CREATED).body(
                ApiResponse.success(Map.of("id", id), "Sifariş yaradıldı"));
    }

    @GetMapping("/{id}")
    public ApiResponse<OrderDto> show(@PathVariable UUID id) {
        return ApiResponse.success(queryBus.ask(new GetOrderQuery(id)));
    }

    @GetMapping("/user/{userId}")
    public ApiResponse<List<OrderDto>> listByUser(@PathVariable UUID userId) {
        return ApiResponse.success(queryBus.ask(new ListOrdersQuery(userId)));
    }

    @PostMapping("/{id}/cancel")
    public ApiResponse<Void> cancel(@PathVariable UUID id, @RequestBody(required = false) CancelBody body) {
        commandBus.dispatch(new CancelOrderCommand(id, body != null ? body.reason() : null));
        return ApiResponse.success(null, "Sifariş ləğv edildi");
    }

    @PatchMapping("/{id}/status")
    public ApiResponse<Void> updateStatus(@PathVariable UUID id, @RequestBody StatusBody body) {
        commandBus.dispatch(new UpdateOrderStatusCommand(id, body.target()));
        return ApiResponse.success(null, "Status yeniləndi");
    }

    public record CancelBody(String reason) {}
    public record StatusBody(OrderStatusEnum target) {}
}
